aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,298 @@
+//       This is because, for CUDA, we will be using Philox4_32_10 engine
+//       inside a kernel and the number of random samples we would need is known
+//       before hand. For the CPU side, it's easier to let CPUGenerator encapsulate
+//       Philox4_32_10 and manage its state. Hence, for CUDA, operator() takes
+//       the Philox4_32_10 object directly and CPU take CPUGenerator object directly.
+namespace at {
+
+// Constants for uniform distribution
+constexpr float POW_2_32_INV = 1.0f/std::numeric_limits<uint32_t>::max();
+constexpr double POW_2_64_INV = 1.0/std::numeric_limits<uint64_t>::max();
+
+/**
+ * Samples a uniform distribution in the range [0,1) of type T
+ */
+template <typename T>
+struct uniform_real_distribution {
+
+  C10_HOST_DEVICE inline uniform_real_distribution(T a_in, T b_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(a_in <= b_in);
+        assert(b_in-a_in <= std::numeric_limits<T>::max());
+      #else
+        AT_ASSERT(a_in <= b_in);
+        AT_ASSERT(b_in-a_in <= std::numeric_limits<T>::max());
+      #endif
+    #endif
+    a = a_in;
+    b = b_in;
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, double>::value, double>::type
+  operator()(at::CPUGenerator* generator){
+    double x = generator->random64() * POW_2_64_INV;
+    if (x == 1.0) {
+      x = std::nextafter(1.0, 0.0);
+    }
+    return (x * (b - a) + a);
+
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, float>::value
+                                          || std::is_same<U, at::Half>::value, float>::type
+  operator()(at::CPUGenerator* generator){
+    float x = generator->random() * POW_2_32_INV;
+    if (x == 1.0f) {
+      x = std::nextafter(1.0f, 0.0f);
+    }
+    return (x * (b - a) + a);
+  }
+
+  private:
+    T a;
+    T b; 
+};
+
+/**
+ * Samples a normal distribution using the Box-Muller method
+ * Takes mean and standard deviation as inputs
+ * Outputs two samples at a time
+ */
+template <typename T>
+struct normal_distribution {
+
+  C10_HOST_DEVICE inline normal_distribution(T mean_in, T stdv_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(stdv_in > 0);
+      #else
+        AT_ASSERT(stdv_in > 0);
+      #endif
+    #endif
+    mean = mean_in;
+    stdv = stdv_in;
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, double>::value, DOUBLE2>::type
+  operator()(at::CPUGenerator* generator){
+    uniform_real_distribution<double> uniform(0.0, 1.0);
+    DOUBLE2 result;
+    double u2 = uniform(generator);
+    double u1 = uniform(generator);
+    // extra pre-caution to make sure log never gets zero
+    double log_val = 1.0 - u1;
+    if (log_val == 0.0) {
+      log_val = std::numeric_limits<double>::min();
+    }
+    double r = ::sqrt(-2.0 * ::log(log_val));
+    result[0] = r * ::cos(2.0 * M_PI * u2) * stdv + mean;
```

Comment:
Might be not a big deal but still:
const double t = 2.0 * M_PI * u2;
.... cos(t) ....
.... sin(t) ....